code/core/src/phone/java/com/adobe/marketing/mobile/services/ui/notification/PushTemplateImageUtil.kt

```diff
@@ -0,0 +1,240 @@
+            } finally {
+                connection?.disconnect()
+            }
+            return bitmap
+        }
+    }
+
+    /**
+     * Downloads an image using the provided uri `String`. Prior to downloading, the image uri
+     * is used to retrieve a [CacheResult] containing a previously cached image. If no cache
+     * result is returned, a call to [download] is made to download then cache the image.
+     *
+     * If a valid cache result is returned then no image is downloaded. Instead, a `Bitmap`
+     * is created from the cache result and returned by this method.
+     *
+     * @param cacheService the AEPSDK [CacheService] to use for caching or retrieving
+     * downloaded image assets
+     * @param uri [String] containing an image asset url
+     * @return [Bitmap] containing the image referenced by the `String` uri
+     */
+    internal fun downloadImage(cacheService: CacheService, uri: String?): Bitmap? {
+        val assetCacheLocation = getAssetCacheLocation()
+        if (assetCacheLocation.isNullOrEmpty() || uri.isNullOrEmpty()) {
+            return null
+        }
+        val cacheResult = cacheService[assetCacheLocation, uri]
+        if (cacheResult != null) {
+            Log.trace(PushTemplateConstants.LOG_TAG, SELF_TAG, "Found cached image for $uri")
+            return BitmapFactory.decodeStream(cacheResult.data)
+        }
+        if (!UrlUtils.isValidUrl(uri)) {
+            return null
+        }
+        val image = download(uri) ?: return null
+        Log.trace(
+            PushTemplateConstants.LOG_TAG,
+            SELF_TAG,
+            "Successfully download image from $uri"
+        )
+        // scale down the bitmap to 300dp x 200dp as we don't want to use a full
+        // size image due to memory constraints
+        val pushImage = scaleBitmap(image)
+        // write bitmap to cache
+        try {
+            bitmapToInputStream(pushImage).use { bitmapInputStream ->
+                cacheBitmapInputStream(
+                    cacheService,
+                    bitmapInputStream,
+                    uri
+                )
+            }
+        } catch (exception: IOException) {
+            Log.trace(
+                PushTemplateConstants.LOG_TAG,
+                SELF_TAG,
+                "Exception occurred creating an input stream from a bitmap: ${exception.localizedMessage}."
+            )
+        }
+        return pushImage
+    }
+
+    /**
+     * Downloads an image using the provided uri `String`. A [Future] task is created to download
+     * the image using a [DownloadImageCallable]. The task is submitted to an [ExecutorService].
+     * @param url [String] containing the image url to download
+     * @return [Bitmap] containing the downloaded image
+     */
+    internal fun download(url: String?): Bitmap? {
+        var bitmap: Bitmap? = null
+        val executorService = executor
+        val downloadTask = executorService.submit(DownloadImageCallable(url))
```

Comment:
Currently download logic waits for each image to download based on the custom network request crafted by this class.  Let us use `NetworkService` to download all the images at once, and wait for a fixed time to see if we have minimum number of images for displaying.

Reply:
Will address in a subsequent PR